packages/output/src/browser/output-widget.tsx

```diff
@@ -14,137 +14,238 @@
+            MessageLoop.sendMessage(widget, Widget.ResizeMessage.UnknownSize);
+        }
     }
 
-    protected render(): React.ReactNode {
-        return <React.Fragment>
-            {this.renderChannelContents()}
-        </React.Fragment>;
+    get onStateChanged(): Event<OutputWidget.State> {
+        return this.onStateChangedEmitter.event;
     }
 
     clear(): void {
-        if (this.outputChannelManager.selectedChannel) {
-            this.outputChannelManager.selectedChannel.clear();
+        if (this.selectedChannel) {
+            this.selectedChannel.clear();
         }
     }
 
     selectAll(): void {
-        if (this.outputChannelManager.selectedChannel) {
-            const element = document.getElementById(OutputWidget.IDs.CONTENTS);
-            if (element) {
-                const selectElementContent = (htmlElement: HTMLElement) => {
-                    const selection = window.getSelection();
-                    const range = document.createRange();
-                    if (selection && range) {
-                        range.selectNodeContents(htmlElement);
-                        selection.removeAllRanges();
-                        selection.addRange(range);
-                    }
-                };
-                element.focus();
-                selectElementContent(element);
+        const editor = this.editor;
+        if (editor) {
+            const model = editor.getControl().getModel();
+            if (model) {
+                const endLine = model.getLineCount();
+                const endCharacter = model.getLineMaxColumn(endLine);
+                editor.getControl().setSelection(new monaco.Range(1, 1, endLine, endCharacter));
             }
         }
     }
 
-    protected renderChannelContents(): React.ReactNode {
-        return <div id={OutputWidget.IDs.CONTENTS}>{this.renderLines()}</div>;
+    lock(): void {
+        this.state = { ...deepClone(this.state), locked: true };
     }
 
-    protected renderLines(): React.ReactNode[] {
-        let id = 0;
-        const result = [];
+    unlock(): void {
+        this.state = { ...deepClone(this.state), locked: false };
+    }
 
-        if (this.outputChannelManager.selectedChannel) {
-            for (const outputChannelLine of this.outputChannelManager.selectedChannel.getLines()) {
-                const lines = outputChannelLine.text.split(/[\n\r]+/);
-                let className;
-                if (outputChannelLine.severity === OutputChannelSeverity.Error) {
-                    className = 'theia-output-error';
-                } else if (outputChannelLine.severity === OutputChannelSeverity.Warning) {
-                    className = 'theia-output-warning';
-                }
-                for (const line of lines) {
-                    result.push(<div className={`theia-output-lines ${className}`} key={id++}>{line}</div>);
-                }
+    get isLocked(): boolean {
+        return !!this.state.locked;
```

Comment:
VS Code calls it `scrollLock`, I wanted to keep it aligned with the _lock_/_unlock_ verbs but I do not mind changing if the community requests this. Otherwise, I am going to leave as is.